### PR TITLE
add HW 18 "glovo"

### PR DESCRIPTION
**Pull request by AlexBearSup**

### CODE REVIEW

**Review by sergvetrov — APPROVED** [review]

**Review by abalmasDA — APPROVED** [review]

File: src/main/java/com/example/glovo/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import com.example.glovo.services.ProductService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping("/orders")
+public class OrderController {
+    private final OrderService orderService;
+    private final ProductService productService;
+    @Autowired
```
Comment (mkharkhuta) [inline src/main/java/com/example/glovo/controller/OrderController.java:14]: you can just autowire services, without using Autowired constructs and private final fields

File: src/main/java/com/example/glovo/controller/OrderController.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import com.example.glovo.dto.Order;
+import com.example.glovo.dto.Product;
+import com.example.glovo.services.OrderService;
+import com.example.glovo.services.ProductService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping("/orders")
+public class OrderController {
```
Comment (mkharkhuta) [inline src/main/java/com/example/glovo/controller/OrderController.java:11]: please, reformat code:
Code -> Reformat Code

File: src/main/java/com/example/glovo/controller/OrderController.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    public OrderController(OrderService orderService, ProductService productService) {
+        this.orderService = orderService;
+        this.productService = productService;
+    }
+    @GetMapping("/{orderId}")
+    public Order get(@PathVariable Integer orderId) {
+        return orderService.get(orderId);
+    }
+    @PostMapping
+    public Integer create(@RequestBody Order order) {
```
Comment (mkharkhuta) [inline src/main/java/com/example/glovo/controller/OrderController.java:24]: what if order with specific id will be created?

File: src/main/java/com/example/glovo/controller/OrderController.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        return orderService.create(order);
+    }
+    @PutMapping("/{orderId}")
+    public void update(@PathVariable Integer orderId, @RequestBody Order updatedOrder) {
```
Comment (mkharkhuta) [inline src/main/java/com/example/glovo/controller/OrderController.java:28]: it's better to return updated data

File: src/main/java/com/example/glovo/controller/OrderController.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        orderService.update(orderId, updatedOrder);
+    }
+    @PatchMapping("/{orderId}/addProduct")
+    public void addProduct(@PathVariable Integer orderId, @RequestBody Product product) {
```
Comment (mkharkhuta) [inline src/main/java/com/example/glovo/controller/OrderController.java:32]: the same, return updated data

File: src/main/java/com/example/glovo/services/OrderService.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    }
+    public int create(Order order) {
+        int orderId = orderIdCounter++;
+        order.setId(orderId);
+        orders.put(orderId, order);
+        return orderId;
+    }
+    public void update(int orderId, Order updatedOrder) {
+        if (orders.containsKey(orderId)) {
+            orders.put(orderId, updatedOrder);
```
Comment (mkharkhuta) [inline src/main/java/com/example/glovo/services/OrderService.java:23]: is it ok that we can map the order with id 1 to key 2?

File: src/main/java/com/example/glovo/services/ProductService.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+import com.example.glovo.dto.Order;
+import com.example.glovo.dto.Product;
+import org.springframework.stereotype.Component;
+import org.springframework.stereotype.Service;
+
+
+@Service
+public class ProductService {
+    public void addToOrder(Order order, Product product) {
```
Comment (mkharkhuta) [inline src/main/java/com/example/glovo/services/ProductService.java:11]: It;s better to move this method to OrderService as actually the order is updating

File: src/main/java/com/example/glovo/services/ProductService.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+        order.getProducts().add(product);
+    }
+    public void removeFromOrder(Order order, Integer productId) {
+        order.getProducts().removeIf(product -> product.getId().equals(productId));
```
Comment (mkharkhuta) [inline src/main/java/com/example/glovo/services/ProductService.java:15]: NPE wil accure if order is null

**Review by mkharkhuta — COMMENTED** [review]
Please check comments. Also it would be nice to add tests

**Pull request merged**